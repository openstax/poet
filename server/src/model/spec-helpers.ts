import expect from 'expect'
import SinonRoot from 'sinon'
import { readFileSync } from 'fs'
import * as path from 'path'
import I from 'immutable'
import { Bundle } from './bundle'
import { Fileish, ValidationKind } from './fileish'

export const REPO_ROOT = path.join(__dirname, '..', '..', '..')

export const read = (filePath: string) => readFileSync(filePath, 'utf-8')

/* Copy/Pasted from ./utils.ts to remove cyclic dependency */
interface PathHelper<T> {
  join: (root: T, ...components: string[]) => T
  dirname: (p: T) => T
  canonicalize: (p: T) => T
}

export const FS_PATH_HELPER: PathHelper<string> = {
  join: path.join,
  dirname: path.dirname,
  canonicalize: (x) => x
}

export function first<T>(col: I.Set<T> | I.List<T>) {
  const f = col.toArray()[0]
  expect(f).toBeTruthy()
  return f
}

export const makeBundle = () => new Bundle(FS_PATH_HELPER, REPO_ROOT)

export function loadSuccess<T extends Fileish>(n: T, skipInitialLoadedCheck = false, expectedErrorCount = 0) {
  if (!skipInitialLoadedCheck) expect(n.isLoaded).toBeFalsy()
  n.load(read(n.absPath))
  expect(n.isLoaded).toBeTruthy()
  expect(n.exists).toBeTruthy()
  expect(n.validationErrors.errors.size).toBe(expectedErrorCount)
  return n // for daisy-chaining
}

export function ignoreConsoleWarnings(fn: () => void) {
  const warnStub = SinonRoot.stub(console, 'warn')
  fn()
  warnStub.restore()
}

export function expectErrors<T extends Fileish>(node: T, validationKinds: ValidationKind[]) {
  const v = node.validationErrors
  expect(v.nodesToLoad.size).toBe(0) // Everything should have loaded
  expect(v.errors.toArray().map(e => e.message).sort()).toEqual(validationKinds.map(v => v.title).sort())
}
