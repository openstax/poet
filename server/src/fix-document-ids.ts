import fs from 'fs'
import { DOMParser, XMLSerializer } from 'xmldom'
import { URI } from 'vscode-uri'
import { ELEMENT_TO_PREFIX, PageNode } from './model/page'
import { expectValue, select } from './model/utils'

const ID_PADDING_CHARS = 5

export function padLeft(text: string, padChar: string, size: number): string {
  if (text.length < size) {
    return (String(padChar).repeat(size) + text).substr((size * -1), size)
  } else {
    return text
  }
}

function buildId(tag: string, counter: number): string {
  const prefix = expectValue(ELEMENT_TO_PREFIX.get(tag), 'BUG: Element was not in the id-prefix map')
  return `${prefix}-${padLeft(String(counter), '0', ID_PADDING_CHARS)}`
}

// Do not add ids to <term> inside a definition.
function termSpecificSelector(e: string): string {
  return e === 'term' ? '[not(parent::cnxml:definition)]' : ''
}

export function fixDocument(doc: Document): void {
  const elsWithIds = select('//cnxml:*[@id]', doc) as Element[]
  const ids = new Set(elsWithIds.map(el => el.getAttribute('id')))
  const xpath = Array.from(ELEMENT_TO_PREFIX.keys()).map(e => `//cnxml:${e}[not(@id)]${termSpecificSelector(e)}`).join('|')
  const els = select(xpath, doc) as Element[]
  const cacheHighId: { [tag: string]: number } = {}
  for (const el of els) {
    const tag = el.tagName.toLowerCase()
    let counter = cacheHighId[tag] > 0 ? cacheHighId[tag] + 1 : 1
    while (ids.has(buildId(tag, counter))) {
      counter++
    }
    ids.add(buildId(tag, counter)) // avoid reusage of new generated id
    cacheHighId[tag] = counter // cache new highest counter
    el.setAttribute('id', buildId(tag, counter))
  }
}

export async function fixModule(p: PageNode): Promise<void> {
  const fsPath = URI.parse(p.absPath).fsPath
  // 3 steps necessary for element id creation: check, fix, save
  // == check xml ==
  // TODO: use cached book-bundle doc data in future for performance increase?
  const data = await fs.promises.readFile(fsPath, { encoding: 'utf-8' })
  const doc = new DOMParser().parseFromString(data)
  // == fix xml ==
  fixDocument(doc)
  // == save xml ==
  const out = new XMLSerializer().serializeToString(doc)
  await fs.promises.writeFile(fsPath, out, { encoding: 'utf-8' })
}
