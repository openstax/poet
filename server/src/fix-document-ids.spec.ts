import { expect } from '@jest/globals'
import assert from 'assert'
import * as xpath from 'xpath-ts'
import { DOMParser, XMLSerializer } from 'xmldom'
import { fixDocument, idFixer, padLeft } from './fix-document-ids'
import mockfs from 'mock-fs'
import { bundleMaker, pageMaker } from './model/spec-helpers.spec'

describe('Element ID creation', () => {
  describe('fixDocument', () => {
    it('check if ids are created', () => {
      const simple = `<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <para id="test">Test Introduction</para>
              <para>no id here</para>
            </content>
          </document>`
      const doc = new DOMParser().parseFromString(simple)
      fixDocument(doc)
      const out = new XMLSerializer().serializeToString(doc)
      expect(out).toMatchSnapshot()
    })
    it('check if ids are created in right order', () => {
      const paraPartialId = `<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <section>
                <para id="para-00001">Test Introduction</para>
                <para>no id here</para>
              </section>
              <section id="sect-00001">
                <para id="para-00002">oh we have id2 here</para>
                <para id="para-00003">oh we have id3 here</para>
                <para>no id here</para>
                <para>no id here</para>
              </section>
            </content>
          </document>`
      const doc = new DOMParser().parseFromString(paraPartialId)
      fixDocument(doc)
      const NS_CNXML = 'http://cnx.rice.edu/cnxml'
      const select = xpath.useNamespaces({ cnxml: NS_CNXML })
      const fixParaNodes = select('//cnxml:para', doc) as Element[]
      assert.strictEqual(fixParaNodes[0].getAttribute('id'), 'para-00001')
      assert.strictEqual(fixParaNodes[1].getAttribute('id'), 'para-00004')
      assert.strictEqual(fixParaNodes[2].getAttribute('id'), 'para-00002')
      assert.strictEqual(fixParaNodes[3].getAttribute('id'), 'para-00003')
      assert.strictEqual(fixParaNodes[4].getAttribute('id'), 'para-00005')
      assert.strictEqual(fixParaNodes[5].getAttribute('id'), 'para-00006')
      const fixSectionNodes = select('//cnxml:section', doc) as Element[]
      assert.strictEqual(fixSectionNodes[0].getAttribute('id'), 'sect-00002')
      assert.strictEqual(fixSectionNodes[1].getAttribute('id'), 'sect-00001')
    })
    it('check if ids are generated with right (short) prefix', () => {
      const xml = `<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <para>no id here</para>
              <equation/>
              <list/>
              <section/>
              <problem/>
              <solution/>
              <exercise/>
              <example/>
              <figure/>
              <definition/>
              <para>
                <term>hello</term>
              </para>
              <table/>
              <quote/>
              <note/>
              <footnote/>
              <cite/>
            </content>
          </document>`
      const doc = new DOMParser().parseFromString(xml)
      fixDocument(doc)
      const out = new XMLSerializer().serializeToString(doc)
      expect(out).toMatchSnapshot()
    })
    it('check if term in defintion does not get an id', () => {
      const xml = `<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <para>
                <term id="term-00001">hello</term>
              </para>
              <definition>
                <term>I should not get an id</term>
              </definition>
              <para>
                <term>need id</term>
              </para>
            </content>
          </document>`
      const doc = new DOMParser().parseFromString(xml)
      fixDocument(doc)
      const out = new XMLSerializer().serializeToString(doc)
      expect(out).toMatchSnapshot()
    })
    it('check high id number generation works right', () => {
      assert.strictEqual(padLeft('1', '0', 5), '00001')
      assert.strictEqual(padLeft('1000', '0', 5), '01000')
      assert.strictEqual(padLeft('10000', '0', 5), '10000')
      assert.strictEqual(padLeft('100000', '0', 5), '100000')
      assert.strictEqual(padLeft('34262934876', '0', 5), '34262934876')
      assert.strictEqual(padLeft('99999', '0', 5), '99999')
      assert.strictEqual(padLeft('9999', '0', 5), '09999')
    })
  })
  describe('fixModule', () => {
    beforeEach(() => {
      mockfs({
        'META-INF/books.xml': bundleMaker({}),
        'modules/m2468/index.cnxml': pageMaker({ extraCnxml: '<para>should get autogenerated id added</para>' })
      })
    })
    afterEach(() => { mockfs.restore() })
    it('fixes a page in-memory', () => {
      const input = pageMaker({ extraCnxml: '<para>should get autogenerated id added</para>' })
      const out = idFixer(input, 'somefilenamedoesnotmatter')
      expect(out).toEqual(expect.stringContaining('<para id="para-00001">'))
    })
  })
})
