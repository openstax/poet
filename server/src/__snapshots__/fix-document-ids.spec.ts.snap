// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element ID creation fixDocument check if ids are created 1`] = `
"<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <para id="test">Test Introduction</para>
              <para id="para-00001">no id here</para>
            </content>
          </document>"
`;

exports[`Element ID creation fixDocument check if ids are generated with right (short) prefix 1`] = `
"<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <para id="para-00001">no id here</para>
              <equation id="eq-00001"/>
              <list id="list-00001"/>
              <section id="sect-00001"/>
              <problem id="prob-00001"/>
              <solution id="sol-00001"/>
              <exercise id="exer-00001"/>
              <example id="exam-00001"/>
              <figure id="fig-00001"/>
              <definition id="def-00001"/>
              <para id="para-00002">
                <term id="term-00001">hello</term>
              </para>
              <table id="table-00001"/>
              <quote id="quote-00001"/>
              <note id="note-00001"/>
              <footnote id="foot-00001"/>
              <cite id="cite-00001"/>
            </content>
          </document>"
`;

exports[`Element ID creation fixDocument check if term in defintion does not get an id 1`] = `
"<document id="new" cnxml-version="0.7" module-id="" xmlns="http://cnx.rice.edu/cnxml" class="introduction">
            <title>Introduction</title>
            <metadata mdml-version="0.5" xmlns:md="http://cnx.rice.edu/mdml">
              <md:uuid>00000000-0000-0000-0000-000000000000</md:uuid>
            </metadata>
            <content>
              <para id="para-00001">
                <term id="term-00001">hello</term>
              </para>
              <definition id="def-00001">
                <term>I should not get an id</term>
              </definition>
              <para id="para-00002">
                <term id="term-00002">need id</term>
              </para>
            </content>
          </document>"
`;
